### ISSUE TITLE
feat:  strip authentication secrets from  private index/download URL

### ISSUE THREAD

**Issue by fkleon** [CycloneDX/cyclonedx-python#646]
Version 4.0.0 includes additional metadata in the SBOM, one of which is the package url.

When using `cyclonedx-py` with the  `--index-url` parameter, the index URL ends up in the component metadata section of the SBOM (this happen with version `v4.0.0-rc.4`)

This can be an issue if a private PyPI registry URL is used that contains authentication credentials, e.g. the [GitLab PyPi repository](https://docs.gitlab.com/ee/user/packages/pypi_repository/#install-a-pypi-package), where the URL looks like:

```
https://<personal_access_token_name>:<personal_access_token>@gitlab.example.com/api/v4/projects/<project_id>/packages/pypi/simple
```

The URL should probably be sanitised first to strip the authentication part; or at least a warning printed for the user?

**Comment by jkowalleck** [comment]
caused by https://github.com/CycloneDX/cyclonedx-python/pull/605

**Comment by jkowalleck** [comment]
thanks for testing the v4.0.0-rc.4.

a possible solution could be, to replacce the password with `*****`? would this be sufficient, @fkleon ?

the mentioned issue affects the following
- all (extra) index urls for
  -  `pipenv`
  - `requirements`
  - `poetry`
- all direct download urls for  
  - `pyproject.toml`
  - `environment` 
  -  `pipenv`
  - `requirements`
  - `poetry`
- all external refererences for 
  - `pyproject.toml`
  - `environment` 
  -  `pipenv`
  - `requirements`
  - `poetry`

**Comment by fkleon** [comment]
HI @jkowalleck! Masking the password is sufficient to avoid accidentally leaking the credentials.

However I'm not sure what the primary purpose of the `url` field is. If those URLs should be comparable across SBOM versions it's probably best to fully strip the credentials (e.g. when running in CI the username/password might be different every time).

So that a given URL: `https://gitlab-ci:SOME-TOKEN@gitlab.example.com/api/v4/projects/some-project/packages/pypi/simple`
Would be normalised to: `https://gitlab.example.com/api/v4/projects/some-project/packages/pypi/simple`

**Comment by jkowalleck** [comment]
> So that a given URL: `https://gitlab-ci:SOME-TOKEN@gitlab.example.com/api/v4/projects/some-project/packages/pypi/simple`
> Would be normalised to: `https://gitlab.example.com/api/v4/projects/some-project/packages/pypi/simple`

will go with this solution. implementation in progress #647

**Comment by jkowalleck** [comment]
a fix was published in version `4.0.0rc5`

**Comment by fkleon** [comment]
Thanks for the quick turnaround. Seems to be working nicely now with `rc5`.

Just wanting to say that the rest of 4.0.0 also looks really good, it's great having the additional metadata (e.g. licenses) and the dependency tree included in the SBOM.
